lib/std/segmented_list.zig

```diff
@@ -0,0 +1,472 @@
+const std = @import("std.zig");
+const assert = std.debug.assert;
+const testing = std.testing;
+const Allocator = std.mem.Allocator;
+
+// Imagine that `fn at(self: *Self, index: usize) &T` is a customer asking for a box
+// from a warehouse, based on a flat array, boxes ordered from 0 to N - 1.
+// But the warehouse actually stores boxes in shelves of increasing powers of 2 sizes.
+// So when the customer requests a box index, we have to translate it to shelf index
+// and box index within that shelf. Illustration:
+//
+// customer indexes:
+// shelf 0:  0
+// shelf 1:  1  2
+// shelf 2:  3  4  5  6
+// shelf 3:  7  8  9 10 11 12 13 14
+// shelf 4: 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30
+// shelf 5: 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62
+// ...
+//
+// warehouse indexes:
+// shelf 0:  0
+// shelf 1:  0  1
+// shelf 2:  0  1  2  3
+// shelf 3:  0  1  2  3  4  5  6  7
+// shelf 4:  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15
+// shelf 5:  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31
+// ...
+//
+// With this arrangement, here are the equations to get the shelf index and
+// box index based on customer box index:
+//
+// shelf_index = floor(log2(customer_index + 1))
+// shelf_count = ceil(log2(box_count + 1))
+// box_index = customer_index + 1 - 2 ** shelf
+// shelf_size = 2 ** shelf_index
+//
+// Now we complicate it a little bit further by adding a preallocated shelf, which must be
+// a power of 2:
+// prealloc=4
+//
+// customer indexes:
+// prealloc:  0  1  2  3
+//  shelf 0:  4  5  6  7  8  9 10 11
+//  shelf 1: 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27
+//  shelf 2: 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59
+// ...
+//
+// warehouse indexes:
+// prealloc:  0  1  2  3
```

Comment:
This sounds really similar to `std::hive` which has been proposed for standardization in C++ for some while now.
It might be interesting to have a [look](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0447r15.html) at it.

In a nutshell, `std::hive` adds a skiplist that enables one to reuse arbitrary fields. I am not sure whether this is really useful in the context of a compiler because I believe most decls to live over the course of the whole compilation.

That said, it is also a variant of `std::deque` with variable block sizes. There is a considerably increased complexity from that. Did you consider using fixed length blocks which would simplify the implementation considerably?

Reply:
I had a look. I think this SegmentedList implementation with power-of-two increasing segment sizes is nicer because unlike the C++ hive data structure, this has O(1) indexing. Also it has amortized O(1) append, same as ArrayList, unlike the C++ hive data structure. Also the C++ hive tries to incorporate the idea of erasing elements, which we don't actually need here.

Overall I'm happy with the simplicity sweet spot that I've found with SegmentedList.